File: Luke Shillabeer/week-7/day-1/anagrammr/anagrammr.rb
def anagram_detect(test_word, list)
  
  # use letters in word sorted alphabetically as key (alpha-key), associate
  # word with its alpha-key in a list (appending to list if multiple words
  # share a key)
  uniqued_words = Hash.new { |h, k| h[k] = [] }
  for word in list
    uniqued_words[word.downcase.chars.sort.join()] << word
  end

  p uniqued_words

  # uniqueify the input-word, check if in hash. Return list of words with 
  # particular alpha-key (removed the test_word from the returned result)
  key_str = test_word.chars.sort.join()
  if !uniqued_words[key_str].nil?
    uniqued_words[key_str] - [test_word]
  end
end

anagram_detect("listen", ["Listen" ,"enlists", "google", "inlets", "banana", "listne"])

File: Luke Shillabeer/week-7/day-1/anagrammr/anagrammr.js
// http://stackoverflow.com/a/3955096/2355035
Array.prototype.remove = function() {
  var what, a = arguments, L = a.length, ax;
  while (L && this.length) {
    what = a[--L];
    while ((ax = this.indexOf(what)) !== -1) {
      this.splice(ax, 1);
    }
  }
  return this;
};

function anagram_detect(test_word, word_list) {
  
  // build an object that contains the uniqu'ed (alphabetically sorted characters)
  // words as keys, words generating that key as list of values
  var unique_words = {};
  for (i = 0; i < word_list.length; i++) {
    sort_word = word_list[i].split('').sort().join('');
    if (unique_words.hasOwnProperty(sort_word)) {
      unique_words[sort_word].push(word_list[i]);
    } else {
      unique_words[sort_word] = [word_list[i]];
    }
  }

  // uniqueify the input-word, check if in object. Return array without
  // the test_word removed from possible list of results
  key_test_word = test_word.split('').sort().join('');
  if (unique_words.hasOwnProperty(key_test_word)) {
    return unique_words[key_test_word].remove(test_word);
  }
}

x = anagram_detect("listen", ["enlists", "google", "inlets", "banana", "listne"]);
console.log(x);
